aws/resource_aws_launch_template.go

```diff
@@ -0,0 +1,976 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+const awsSpotInstanceTimeLayout = "2006-01-02T15:04:05Z"
+
+func resourceAwsLaunchTemplate() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsLaunchTemplateCreate,
+		Read:   resourceAwsLaunchTemplateRead,
+		Update: resourceAwsLaunchTemplateUpdate,
+		Delete: resourceAwsLaunchTemplateDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"name_prefix"},
+				ValidateFunc:  validateLaunchTemplateName,
+			},
+
+			"name_prefix": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateLaunchTemplateName,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+					if len(value) > 255 {
+						errors = append(errors, fmt.Errorf(
+							"%q cannot be longer than 255 characters", k))
+					}
+					return
+				},
+			},
+
+			"client_token": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"default_version": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+
+			"latest_version": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+
+			"block_device_mappings": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"device_name": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"no_device": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"virtual_name": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"ebs": {
+							Type:     schema.TypeSet,
```

Comment:
To simplify the implementation and future maintenance, we can use `schema.TypeList` here 👍